x/ccv/provider/keeper/keeper.go

```diff
@@ -1324,3 +1324,25 @@ func (k Keeper) GetToBeOptedOut(
 
 	return addresses
 }
+
+func (k Keeper) IdentifyConsumerChainIDFromIBCPacket(ctx sdk.Context, packet channeltypes.Packet) (string, error) {
+	channel, ok := k.channelKeeper.GetChannel(ctx, packet.DestinationPort, packet.DestinationChannel)
+	if !ok {
+		return "", errorsmod.Wrapf(channeltypes.ErrChannelNotFound, "channel not found for channel ID: %s", packet.DestinationChannel)
+	}
+	if len(channel.ConnectionHops) != 1 {
+		return "", errorsmod.Wrap(channeltypes.ErrTooManyConnectionHops, "must have direct connection to consumer chain")
+	}
+	connectionID := channel.ConnectionHops[0]
+	_, tmClient, err := k.getUnderlyingClient(ctx, connectionID)
+	if err != nil {
+		return "", err
+	}
+
+	chainID := tmClient.ChainId
+	if _, ok := k.GetChainToChannel(ctx, chainID); !ok {
```

Comment:
It might make sense to have another check here that the channel returned by `k.GetChainToChannel` is the same as the channel that was used to send the packet?
It seems to me that with the current approach someone could send an IBC packet from a non-consumer chain that happens to have the same `chainID` as a consumer chain and this would still return a `chainID`.